File: CODING/SORTINGS/Insertion_Sort.c
/* 
Insertion Sort

Time Complexity: O(n ^ 2) 
Space Complexity : O(1)
*/
#include<stdio.h>
#include<stdlib.h>

void PrintArray(int *arr, int size) {

	printf("Array : ");

	for(int i = 0; i < size; ++i) {

		printf(" %d,", arr[i]);
	}
}

void Sort(int *arr, int size) {

	int i, j, key;
	// picking up all the elements
	for(j = 1; j < size; ++j) {

		key = arr[j];
		i = j - 1;
		/* Move elements of arr[0..i-1], that are
          	greater than key, to one position ahead
          	of their current position */
		while(i >= 0 && arr[i] > key) {

			arr[i + 1] = arr[i];
			--i;
		}
		arr[i + 1] = key;
	}
}

int main() {

	int *arr, size;

	printf("Enter size of the Array : ");	
	scanf("%d", &size);	

	arr = (int*)malloc(sizeof(int) * size);	

	printf("Enter the Array : ");
	for(int i = 0; i < size; ++i) {
		scanf("%d", &arr[i]);
	}

	PrintArray(arr, size);
	Sort(arr, size);
	PrintArray(arr, size);
}

File: interview/src/sort/InsertionSort.java
/**
 * Date 06/04/2022
 *
 * @author akhilpathivada
 *
 * Insertion Sort Algorithm
 *
 * Time Complexity : O(N ^ 2)
 * Space Complexity : O(1)
 */
package sort;

import java.util.Arrays;

public class InsertionSort {
        private void sort(int[] arr, int n) {
                int i, j, key;
                // picking up all the elements;
                for (i = 1; i < n; ++i) {
                        key = arr[i];
                        // Move elements of arr[0..i-1], that are greater than key, to one position ahead
                        // of their current position
                        for (j = i - 1; j >= 0 && arr[j] > key; --j) {
                                arr[j + 1] = arr[j];
                        }
                        arr[j + 1] = key;
                }
        }
        public static void main(String args[]) {
                int arr[] = { 12, 11, 13, 5, 6, 7 };
                System.out.println("Given Array" + Arrays.toString(arr));
                new InsertionSort().sort(arr, arr.length);
                System.out.println("Sorted array" +  Arrays.toString(arr));
        }
}
